open_flamingo/src/flamingo_lm.py

```diff
@@ -104,6 +109,43 @@ def init_flamingo(
         )
         self.init_flamingo_layers(gradient_checkpointing)
         self.media_token_id = media_token_id
+
+        # modify the embedding layer to support decoupling
+        input_embed_weights = self.get_input_embeddings().weight
+        self.set_input_embeddings(
+            FlamingoDecoupledEmbedding(
+                num_embeddings=input_embed_weights.shape[0],
+                num_additional_embeddings=new_tokens,
+                embedding_dim=input_embed_weights.shape[1],
+                partially_freeze=True,
+            )
+        )
+        self.get_input_embeddings().weight = input_embed_weights
+
+        # create a get_output_embeddings() / set_output_embeddings() method if it doesn't exist
+        # this is needed for compatibility
+        if not hasattr(self, "get_output_embeddings"):
+            self.get_output_embeddings = lambda: self.lm_head
+            self.set_output_embeddings = lambda x: setattr(self, "lm_head", x)
```

Comment:
Could we move this to factory.py to match how we handle get_input_embeddings for MPT-1B (factory.py:line 73)?  To avoid hard-coding for MPT